File: 04-19 - Grafos/tads/graph/Graph.java
package tads.graph;

import java.util.Iterator;

public interface Graph {
  public void addEdge(int from, int to);

  public void addEdge(int from, int to, int cost);

  public void removeEdge(int from, int to);

  public Iterator<Integer> vertices();

  public Iterator<Integer> adjacents(int v);

  public Iterator<Edge> edges();

  public int vertexCount();

  public int[][] adjacencyMatrix();

  public boolean hasEdge(int from, int to);

  public int getEdgeCost(int from, int to);
}

File: 04-19 - Grafos/tads/graph/graph.cpp
#pragma once

#include "../iterator/iterator.cpp"
#include "edge.cpp"

class Graph
{
public:
  virtual void addEdge(int from, int to) = 0;

  virtual void addEdge(int from, int to, int cost) = 0;

  virtual void removeEdge(int from, int to) = 0;

  virtual Iterator<int> *vertices() = 0;

  virtual Iterator<int> *adjacents(int v) = 0;

  virtual Iterator<Edge> *edges() = 0;

  virtual int vertexCount() = 0;

  virtual int **adjacencyMatrix() = 0;

  virtual bool hasEdge(int from, int to) = 0;

  virtual int getEdgeCost(int from, int to) = 0;
};
